File: adventofcode_hardmode/src/main/kotlin/aoc2019/day06/day6.kt
package aoc2019.day06

import java.io.File

typealias ChildToParent = Map<String, String>

fun parseTree(input: List<String>): ChildToParent =
    input.map { it.split(")") }.associate { (parent, child) -> child to parent }

fun getAncestors(childToParent: Map<String, String>, node: String): Sequence<String> = sequence {
    var parent = childToParent[node]
    while (parent != null) {
        yield(parent)
        parent = childToParent[parent]
    }
}

fun evaluatorOne(childToParent: ChildToParent): Int =
    childToParent.keys.sumOf { getAncestors(childToParent, it).count() }

fun evaluatorTwo(childToParent: ChildToParent): Int {
    val youAncestors = getAncestors(childToParent, "YOU").withIndex().associate { it.value to it.index }
    val sanAncestors = getAncestors(childToParent, "SAN").toList()

    sanAncestors.forEachIndexed { index, ancestor ->
        youAncestors[ancestor]?.let { return it + index }
    }

    error("No common ancestor found between YOU and SAN")
}

fun readLinesFromFile(filePath: String): Result<List<String>> =
    runCatching { File(filePath).readLines() }

fun main() {
    readLinesFromFile("src/main/resources/aoc2019/day06.txt")
        .onSuccess {
            val childToParent = parseTree(it)
            println("Part One: ${evaluatorOne(childToParent)}")
            println("Part Two: ${evaluatorTwo(childToParent)}")
        }
        .onFailure {
            println("Error reading file: ${it.message}")
        }
}

File: 2019/src/main/scala/day06/day6.scala
package day06

import scala.util.{Try, Success, Failure, Using}
import scala.io.Source

type ChildToParent = Map[String, String]

def parseTree(input: List[String]) = input.collect {
    case s"$parent)$child" => child -> parent
}.toMap

def getAncestors(trees: ChildToParent, node: String): List[String] = {
    return Iterator.iterate(Option(node))(_.flatMap(trees.get)).takeWhile(_.isDefined).drop(1).map(_.get).toList
}

def evaluatorOne(trees: ChildToParent): Int = {
    return trees.keysIterator.map(getAncestors(trees, _).size).sum
}

def evaluatorTwo(trees: ChildToParent): Int = {
    val youAncestor = getAncestors(trees, "YOU").toSet
    val sanAncestor = getAncestors(trees, "SAN").toSet

    // finding the symmetric difference between the two sets
    return ((youAncestor &~ sanAncestor) | (sanAncestor &~ youAncestor)).size
}

def readLinesFromFile(filePath: String): Try[List[String]] =
    Using(Source.fromResource(filePath))(_.getLines().toList)

def hello(): Unit = {
    readLinesFromFile("day06.txt") match {
        case Success(lines) => {
            val trees = parseTree(lines)
            println(s"Part One: ${evaluatorOne(trees)}")
            println(s"Part Two: ${evaluatorTwo(trees)}")
        }
        case Failure(exception) => {
            println(s"Error reading file: ${exception.getMessage}")
        }
    }
}